05libraries/sec01ChoosingLibraries.md

```diff
@@ -0,0 +1,188 @@
+
+When you plan to distribute code:
+
+* Don't write your own license, unless you use legal advice and you understand its consequences.
+* Check [GitHub's advice][Chooselicense], and [OSI][OSI] for choosing your own license.
+* Try to pick one of the standard ones for [compatibility][LicesnseCompatibilityWiki].
+
+For an in-depth understanding we recommend you read some works about licenses:
+
+- [Understanding Open Source and Free Software Licensing][LicensingBookWC] ([web][LicensingBook], [UCL][LicensingBookUCL], [pdf][LicensingBookPDF]),
+- [The public domain][PDBook]
+
+**Note**: Once a 3rd party has your code under a license agreement, their restrictions are determined by that version of the code.
+
+### Stability: Is a license stable?
+
+Some libraries are so new their *public API* or *interface* is still subject to change. This is usually signalled by the project being in *alpha* or *beta* stages, either before an initial 1.0.0 release, or before a new major x.0.0 release. Some projects (like Python itself) ensure that all *minor* versions will not intentionally introduce breaking changes (i.e. you can use the same code moving from 3.10 to 3.11) but keep *breaking changes* to new major versions (i.e. moving from Python 2 to Python 3). If you haven't come across this idea, read about [semantic versioning](https://www.geeksforgeeks.org/introduction-semantic-versioning/).
+
+When choosing a library to use with your own project, try to use a *stable* version, i.e. one with a stable interface. 
+
+### Efficiency: Is the library fast enough?
+
+Libraries, particularly the good ones, tend to be well-optimised, that is their algorithms and data structures have been tweaked to get the best performance. For performance-critical libraries (like many used in numerical computations) the library developers should include some details about the performance of the library in its documentation. This is where you should ideally look for information about the performance. Otherwise, try to find comparisons with other, similar libraries to understand the performance.
+
+While you are unlikely to beat a library's performance with a custom algorithm, sometimes custom code can be faster due to a tradeoff between flexibility and performance. If you have already used the library but think you might be able to beat a library's performance:
+
+1. test the performance of the library's implementation
+2. write some unit tests using the library's implementation
+3. write your custom implementation
+4. modify the unit tests to test your implementation
+5. test the performance of your custom code and compare to the library performance
+
+By testing correctness and performance on both the library and your custom code, you can understand whether it's worth it to commit to either.
+
+### Up-to-date: Is the library regularly updated?
+
+Libraries that are not regularly maintained can "rot", that is:
+
+- bugs don't get fixed
+- bugs in dependencies don't get fixed
```

Comment:
Maybe include that libraries which are not maintained can drift from their dependencies, so that they end up becoming incompatible unless you have an old version of that dependency (which may be undesirable if it's a common library that you use elsewhere like a depedency on linear algebra)